pytorch_lightning/callbacks/pruning.py

```diff
@@ -215,175 +217,183 @@ def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
+        """
+
+        def copy_param(new, old, name: str) -> None:
+            dst = getattr(new, name)
+            src = getattr(old, name)
+            if dst is None or src is None or not isinstance(dst, torch.Tensor) or not isinstance(src, torch.Tensor):
+                return
+            dst.data = src.data.to(dst.device)
+
+        for d in self._original_layers.values():
+            copy, names = d["data"], d["names"]
+            if self._resample_parameters and hasattr(copy, "reset_parameters"):
+                copy = deepcopy(copy)  # keep the original parameters
+                copy.reset_parameters()
+            for i, name in names:
+                new, new_name = self._parameters_to_prune[i]
+                copy_param(new, copy, name)
 
     def _apply_local_pruning(self, amount: float):
-        for module, param in self._parameters_to_prune:
-            self.pruning_fn(module, name=param, amount=amount)
+        for module, name in self._parameters_to_prune:
+            self.pruning_fn(module, name=name, amount=amount)
 
     def _resolve_global_kwargs(self, amount: float):
-        kwargs = {}
         self._global_kwargs["amount"] = amount
-        params = inspect.signature(self.pruning_fn).parameters
-        for p_name in params:
-            if p_name != "self":
-                param = self._global_kwargs.get(p_name)
-                if param is not None:
-                    kwargs[p_name] = param
-        return kwargs
+        params = set(inspect.signature(self.pruning_fn).parameters)
+        params.discard("self")
+        return {k: v for k, v in self._global_kwargs.items() if k in params}
 
     def _apply_global_pruning(self, amount: float):
         pytorch_prune.global_unstructured(
             self._parameters_to_prune, pruning_method=self.pruning_fn, **self._resolve_global_kwargs(amount)
         )
 
-    def apply_pruning(self, trainer: 'pl.Trainer', pl_module: LightningModule):
-        amount = self._resolve_amount(trainer.current_epoch)
-
-        # the user could control the pruning frequency with amount_fn
-        if amount == 0 or amount is None:
-            return
+    @staticmethod
+    def _get_pruned_stats(module: nn.Module, name: str) -> Tuple[int, int]:
+        attr = f"{name}_mask"
+        if not hasattr(module, attr):
+            return 0, 1
+        mask = getattr(module, attr)
+        return (mask == 0).sum().item(), mask.numel()
+
+    def apply_pruning(self, amount: Union[int, float]):
+        """ Applies pruning to ``parameters_to_prune``. """
+        if self._verbose:
+            stats = [self._get_pruned_stats(m, n) for m, n in self._parameters_to_prune]
 
         if self._use_global_unstructured:
             self._apply_global_pruning(amount)
         else:
             self._apply_local_pruning(amount)
 
-        if self._use_lottery_ticket_hypothesis:
-            self.apply_lottery_ticket_hypothesis()
+        if self._verbose:
```

Comment:
I think we could have 2 levels of verbose. Overall, and detailed.

Reply:
What's the difference between the two?